common/components/searching.hpp.inc

```diff
@@ -0,0 +1,230 @@
+ * @param length  the length of the partitioned range
+ * @param group   the coalescing group executing the search
+ * @param p  the predicate to be evaluated on the range - it should not have
+ *           side-effects and map from `IndexType` to `bool`
+ * @returns  the index of `middle`, i.e., the partition point
+ */
+template <typename IndexType, typename Group, typename Predicate>
+__device__ IndexType group_wide_search(IndexType offset, IndexType length,
+                                       Group group, Predicate p)
+{
+    // binary search on the group-sized blocks
+    IndexType num_blocks = (length + group.size() - 1) / group.size();
+    auto group_pos = binary_search(IndexType{}, num_blocks, [&](IndexType i) {
+        auto idx = i * group.size();
+        return p(offset + idx);
+    });
+    // case 1: p is true everywhere: middle is at the beginning
+    if (group_pos == 0) {
+        return offset;
+    }
+    /*
+     * case 2: p is false somewhere:
+     *
+     * p(group_pos * g.size()) is true, so either this is the partition point,
+     * or the partition point is one of the g.size() - 1 previous indices.
+     *   |block group_pos-1|
+     * 0 | 0 * * * * * * * | 1
+     *       ^               ^
+     *       we load this range, with the 1 acting as a sentinel for ffs(...)
+     *
+     * additionally, this means that we can't call p out-of-bounds
+     */
+    auto base_idx = (group_pos - 1) * group.size() + 1;
+    auto idx = base_idx + group.thread_rank();
+    auto pos = ffs(group.ballot(idx >= length || p(offset + idx))) - 1;
+    return offset + base_idx + pos;
+}
+
+
+/**
+ * @internal
+ * Generic search that finds the first index where a predicate is true.
+ * It assumes that the predicate partitions the range [offset, offset + length)
+ * into two subranges [offset, middle), [middle, offset + length) such that
+ * the predicate is `false` for all elements in the first range and `true` for
+ * all elements in the second range. `middle` is called the partition point.
+ *
+ * It executes `log2(length) / log2(group.size())` calls to `p` that effectively
+ * follow a random-access pattern.
+ *
+ * This implementation is based on the w-partition search mentioned in
+ * Green et al., "GPU merge path: a GPU merging algorithm"
+ *
+ * @param offset  the starting index of the partitioned range
+ * @param length  the length of the partitioned range
+ * @param group   the coalescing group executing the search
+ * @param p  the predicate to be evaluated on the range - it should not have
+ *           side-effects and map from `IndexType` to `bool`
+ * @returns  the index of `middle`, i.e., the partition point
+ */
+template <typename IndexType, typename Group, typename Predicate>
+__device__ IndexType group_ary_search(IndexType offset, IndexType length,
+                                      Group group, Predicate p)
+{
+    IndexType end = offset + length;
+    // invariant: [offset, offset + length] contains middle
```

Comment:
I'm unsure of the phrasing here. Technically, it's true this is the pivot space (where a pivot could rightfully exist according to the definition), but "offset+length" is of course out of boundaries of the tested space.  Maybe it would be useful to specifically note somewhere that if the predicate isn't satisfied anywhere, then the pivot has to be `offset + length`.

Reply:
I added a clarifying comment to the function documentations.